lib/spack/spack/compiler.py

```diff
@@ -406,6 +407,53 @@ def c11_flag(self):
                                       "the C11 standard",
                                       "c11_flag")
 
+    # Run the compiler to get compiler version info
+    # Use the run environment of the compiler to do so
+    # Note: This is not a class method. The class methods are used to detect
+    # compilers on PATH based systems, and do not set up the run environment of
+    # the compiler. This method can be called on `module` based systems as well
+    def get_real_version(self):
+        """Query the compiler for its version.
+
+        This is the "real" compiler version, regardless of what is in the
+        compilers.yaml file, which the user can change to name their compiler.
+
+        Use the runtime environment of the compiler (modules and environment
+        modifications) to enable the compiler to run properly on any platform.
+        """
+        # store environment to replace later
+        backup_env = os.environ
+
+        # load modules and set env variables
+        for module in self.modules:
+            # On cray, mic-knl module cannot be loaded without cce module
+            # See: https://github.com/spack/spack/issues/3153
+            if os.environ.get("CRAY_CPU_TARGET") == 'mic-knl':
+                spack.util.module_cmd.load_module('cce')
+            spack.util.module_cmd.load_module(module)
+
+        # apply other compiler environment changes
+        env = spack.util.environment.EnvironmentModifications()
+        env.extend(spack.schema.environment.parse(self.environment))
+        env.apply_modifications()
+
+        cc = spack.util.executable.Executable(self.cc)
+        output = cc(self.version_argument,
+                    output=str, error=str,
+                    ignore_errors=tuple(self.ignore_version_errors))
+
+        # restore environment
+        # replacing the os.environ dictionary in python does not effect the
+        # environment variables in process space, so we have to modify the
+        # environment in place
+        for key in set(os.environ.keys()) | set(backup_env.keys()):
+            if key in backup_env:
+                os.environ[key] = backup_env[key]
+            else:
+                del os.environ[key]
```

Comment:
effect -> affect

Also: Instead of this loop, just use `.copy()` above and do:

```
os.environ.clear()
os.environ.update(backup_env)
```

This makes `os.environ == backup_env` in-place (since it's some special kind of `dict`).